SCons/Environment.py

```diff
@@ -1041,11 +1044,17 @@ def get_CacheDir(self):
         except AttributeError:
             path = SCons.Defaults.DefaultEnvironment()._CacheDir_path
         try:
-            if path == self._last_CacheDir_path:
+            if (path == self._last_CacheDir_path
+                and self.get("CACHEDIR_CLASS", SCons.CacheDir.CacheDir) == type(self._last_CacheDir)):
```

Comment:
sider complained about using type instead of isinstance, but I think in this case it should be type check. example use case: you have a compressed cachedir class, and you want to extend it to support encryption so you make another class which uses super calls to complete the compression, and then encrypt.

Reply:
pretty sure [isinstance](https://docs.python.org/3/library/functions.html#isinstance) would be true for `isinstance(child_class,parent_class)`

Reply:
yeah so in my example use case you would want to make sure the encrypted derived class got reinstantiated, so you would not want the early return here. 

In others words the only time we don't reinstanciate is if the path and class match exactly.

Reply:
O.k. Now I understand what you're trying to do.

I think the canonical way to do this would be?
`type(self.get("CACHEDIR_CLASS", SCons.CacheDir.CacheDir)) is type(self._last_CacheDir)`

Since theres only one class, you should use `is` instead of `==` and also you need to compare apples to apples? (type to type)

Reply:
Though not sure if that will make sider happy or not.. but seems more correct?

Reply:
okay sure I'll give that a shot and see what sider thinks

Reply:
sider was still mad, but I found another alternative that makes side happy